Fix multiple issues in `lookup_drop_cname` (supersedes #286)

friendlyhugo opened the pull request:
This PR supersedes #286 (as it cherry-picks @szenker's and @pemensik's commits).

This PR fixes a couple of issues with `lookup_drop_cname`, which cause lookups to be retained even after their corresponding CNAME records are removed. It also corrects the lookup flags comparison in `lookup_drop_cname` which never matched the `flags` value.

At the time, #286 was blocked on #551 being merged - #682, which superseded #551 was merged a couple weeks back.

Thanks to @szenker and @pemensik for their work on #286!

friendlyhugo committed:
avahi-core: fix flags comparison in lookup_find and lookup_drop_cname

Previously avahi would only accept flags if they matched l->flags exactly.
This is incorrect in lookup_drop_cname since it was always called with
flags=0, which caused lookups to linger. This is also incorrect for
lookup_find, where the comparison to b->flags can be against 0, which
means "any transport". This caused lookups to be added incorrectly.

Fixes https://github.com/avahi/avahi/issues/797

friendlyhugo commented on avahi-core/browse.c:
```diff
@@ -155,6 +155,19 @@ static AvahiSRBLookup* lookup_ref(AvahiSRBLookup *l) {
     return l;
 }
 
+static int avahi_lookup_match(AvahiSRBLookup *l, AvahiIfIndex interface, AvahiProtocol protocol, AvahiLookupFlags flags, AvahiKey *key) {
```
Fixed.

friendlyhugo commented:
@evverx, I updated the commit message. WDYT?

friendlyhugo commented:
I've opened #797 to track this.

evverx commented:
Looking at the commit messages they just say what they do. Can you add stuff like "prevents avahi from sending queries when CNAMES expire" to clarify specific issues they address?

friendlyhugo committed:
avahi-core: fix lookup_find and lookup_drop_cname to find the correct AvahiSRBLookup instance in case the protocol member is AVAHI_PROTO_UNSPEC

This typo could lead to the creation of duplicate lookups in lookup_add.

Fixes https://github.com/avahi/avahi/pull/286
Co-authored-by: Sebastian Zenker <sebastian.zenker@gmx.de>
Co-authored-by: Petr Menšík <pemensik@redhat.com>

friendlyhugo commented:
> It turns out with your patch applied the DoS mentioned in #684 is no longer reproducible

Great, I added a "Fixes" line to the commit message.

> Either way can you update the commit messages?

I updated them and squashed the two commits from #286.

evverx commented:
Actually it's still possible to mess up the events and crash avahi-browse by sending another CNAME so "Fixes https://github.com/avahi/avahi/issues/684" can be removed. I'll remove the important label.

evverx commented:
> Anything apart from running the test with flags=0?

The test doesn't actually fail when it's run with `flags=0`. It just emits
```
browse.c: Failed to create SRBLookup.
browse.c: Failed to create SRBLookup.
```
and keeps going so in its current form it can't catch that. I'd need to think about how that part can be covered better.

> Should I add it to this branch or open a new PR?

I think the tests should go along with the code to make sure it actually works so it would be better to add those things here.

friendlyhugo commented:
I decided to add a new test to ensure `lookup_find` functions correctly. It essentially checks whether the final lookup has been added, which, as you've mentioned, would fail before this change.

evverx commented:
> It essentially checks whether the final lookup has been added

I have to say it's a really creative way to make sure that the CNAME pointing to itself isn't added a gazillion times.

I've been looking at 78eab31128479f06e30beb8c1cbf99dd921e2524 and wondering why `lookup_exists_in_path` is needed at all. If `avahi_lookup_match` can look for lookups correctly now `lookup_handle_cname` can detect loops by looking for lookups before adding them. If they already exists it's a loop. Or is there a reason that it can't be handled that way?

evverx commented:
Also with the new test added the code path where `AVAHI_LOOKUPS_PER_BROWSER_MAX` is hit is no longer exercised. I think it should be covered as well by adding a bunch of different CNAMEs or something like that.

evverx commented:
What's about https://github.com/avahi/avahi/commit/78eab31128479f06e30beb8c1cbf99dd921e2524 though? It looks convoluted for no reason. I'm planning to revert https://github.com/avahi/avahi/pull/682 and then reintroduce those things one by one with proper commit messages and links to advisories.

friendlyhugo committed:
avahi-core: move AVAHI_LOOKUPS_PER_BROWSER_MAX to header file

This constant needs to be accessible for smoke-tests.

friendlyhugo committed:
avahi-core: expose remove_entry as avahi_cache_remove_entry

This function needs to be exposed for smoke-tests.

friendlyhugo committed:
smoke-tests: add additional CNAME test cases

These new test cases protect against regressions related to issues #286 and #797.
